### ISSUE TITLE
Reference Target naming

### ISSUE THREAD

**Issue by alice** [WICG/webcomponents#1087]
There has been a reasonable amount of [feedback](https://github.com/w3ctag/design-reviews/issues/961) that the name `referenceTarget` is difficult to understand.

I think it's a reasonable name (ID _references_ to the host should instead go to the designated _target_), but it might be good to see if we can come up with some alternatives.

One thing we should keep in mind is that if we choose to use [separate attributes](https://github.com/WICG/webcomponents/blob/gh-pages/proposals/reference-target-explainer.md#use-separate-attributes-for-each-forwarded-attribute) for each attribute for [Phase 2](https://github.com/WICG/webcomponents/blob/gh-pages/proposals/reference-target-explainer.md#-phase-2-shadowroot-referencetargetmap-attribute), it would be nice to have a noun which can be used in combination with existing attributes (e.g. "shadowrootariaactivedescendant**target**").

**Comment by alice** [comment]
One comparison that gets made a lot is between `referenceTarget` and `delegatesFocus`. 

"Delegates" is a verb in `delegatesFocus`, and `delegatesFocus` is a boolean, so there's no close mapping possible, but we could potentially align with the related spec concept of of [focus delegate](https://html.spec.whatwg.org/#focus-delegate) with something like `referenceDelegate`.

In fact, that was the name originally proposed: https://github.com/WICG/webcomponents/blob/gh-pages/proposals/reference-target-explainer.md#alternative-names-for-the-feature-reference-target

**Comment by rniwa** [comment]
hm... how about `idForwarding` or `idTarget`?

**Comment by Westbrook** [comment]
> hm... how about idForwarding or idTarget?

Feels like anything specifically revolving around “id” flies in the face of W3C TAG wanting to find a path away from needing to rely on IDs to reference things across the DOM. 🤷🏼‍♂️

As a component author, I’ve no problems with `referenceTarget`, particularly in that I don’t need to read from it almost ever and expect to write to it only slightly more frequently. Naming only feels like it would be of issue here if (when?) the suggestion away from a micro DSL for maps in Level 2 is enforced.

Not to make a specific vote for it, but if a rename is required… in a world where `commandFor` is becoming a thing would expanding on “for” in some way be appropriate? e.g. `delegatesFor` or `targetFor`…

**Comment by alice** [comment]
To me `commandFor` is part of a sentence: "button `foo` is a[n open] `commandFor` popover `bar`" -> `<button id="foo" commandfor="bar">`.  I can't quite see how to map that on to what `referenceTarget` is doing - what did you have in mind?

I agree that `id` seems like it might be a bit too specific in the event that we do develop another way to target elements for references. `forwarding` definitely does match my mental model of what the attribute is doing, though, although to me "forwarding" implies a boolean state, rather than a transitive verb ("forward to") or noun ("delegate").

One more pitch for `delegate`: there is a related problem which we deliberately (and correctly) scoped out here, but which we might want to revisit eventually, and that is that we might want to allow authors to ["move" applicable attributes](https://github.com/WICG/webcomponents/issues/1073 ) on the host, like `role`, to a designated element within the shadow root. If we did eventually investigate a parallel API for that capability, I think `delegate` would be a natural fit there.

I do agree that "delegate" does fail the "straightforward English" test, though. It's also easily confused with the verb "deleGATE".

**Comment by Westbrook** [comment]
In the case that we could "move" `role` (or similar) via a `roleDelegate` (or so), it does open the door for next level passing as I've discussed the desire for in https://github.com/WICG/webcomponents/issues/1068. It's almost like we would be drifting towards an `attributeDelegate` at some point. That's CLEARLY not what the API is at this time, however...

While, I was not quite sure how to turn `commandFor` into a usable property/attribute, I was noticing that we start with `<label for="...">` references, a couple others, and now `<button commandfor="...">` references, it seemed there might be something there. If we were "doing more 😞" then it might make sense to see a property/attribute like `attributesfor`. But, I realize that would likely derail much of the work we've done to date, which I am quite remiss to do.

I can definitely get behind a value like `referenceDelegate`! It's much like `referenceTarget`, which I'd be fine with, too, but seems like others might not be? However, language not withstanding, there would be so much gravity for the to become `delegatesReference` if only for the way it would rhyme with existing API.

**Comment by sorvell** [comment]
> There has been a reasonable amount of https://github.com/w3ctag/design-reviews/issues/961 that the name referenceTarget is difficult to understand.

I addition to being hard to understand, there has also [been](https://github.com/w3ctag/design-reviews/issues/961#issuecomment-2457573238) [feedback](https://github.com/w3ctag/design-reviews/issues/961#issuecomment-2483716803) that it's too long.

### Suggestions

| name | understandability | size |
|------|----------------|------|
| `to` | not confusable but similar to `for` need docs to understand | 2 |
| `refsto` | more explicit than `to` and the term `ref` is generally used | 6 |
| `refdelegate` | similar to `referencedelegate` | 11 |
| `delegate` | generic like `to` but longer | 8 |
| `delegatesto` | clear verb (similar to `aria-flowto`) | 11 |

**Comment by alice** [comment]
I quite like `ref` or `refs` as a general direction.

I'll also note some potential prior art in Microsoft's Fluent UI system (React-based) which uses the term "primary slot" for the equivalent of the phase 1 API (plus the "Level 2" ideas Westbrook described): https://github.com/microsoft/fluentui/blob/511c32a4cd38fe82ade4677245b2424b323407fa/rfcs/convergence/native-element-props.md#introduce-the-concept-of-a-primary-slot

**Comment by dandclark** [comment]
This was discussed in the [WebComponentsCG 2025 Q1 Face-to-Face](https://github.com/WICG/webcomponents/issues/1083#issuecomment-2660170704) with `delegatesTo` emerging as the preferred name.

> Dan: we should bikeshed the name - there isn’t a great consensus on the name. One complaint is that it is too long, especially in phase 2, when the attribute gets really long. There is an idea to shorten it - refs, or delegaterefs. Not sure there is anything people have fallen in love with. Status quo would be to run with what we have. If people want to advocate for a different thing, this might be a good time to do that.
Anne van Kesteren: I don’t mind delegate, but I don’t feel particularly strongly.
Olli: I don’t feel strongly, but I also don’t mind delegate.
Mayank: I don’t mind delegate - ref can be an overloaded term, especially with frameworks today. Between delegate and delegatesTo, the latter is more clear.
Dan: I slightly prefer delegates as a term, the length is slightly punishing. On templates it would be `shadowrootdelegatesreferenceariadescribedby`
Anne: can we just use delegates?
Dan: is that confusing with delegatesfocus
Keith Cirkel: could we do then shadowrootdelegatesariadescribedby (that is, remove the reference)
Westbrook: (re delegates vs delegate) delegate is both a verb and a noun
Mayank: should it be delegateId?
Anne: it’s not always an id - it’s the delegate element
Keith: the other id references don’t use id in the name, they use `target`.
\<Some consensus on delegatesTo\>
Dan: would it be valuable to have a survey? If there’s no champion for an alternative that may be worthwhile
Keith: If no one is in objective to delegatesTo, then maybe we don’t need a survey
Dan: I’m okay with having that be the new default pick.
Mayank: is that for both phase 1 and phase 2? It feels confusing for phase 2?
Dan: yeah, potentially we could reword it, to have `to` at the end
Anne: potentially you would remove the `to`
Mayank: `delegatesPopover` does read like a boolean.
Dan: maybe we could sort that out in a phase 2 discussion. It sounds like we can use `delegatesTo`.
Ryosuke: there is a confusion around the implementation around delegatesFocus. Maybe we could have delegates take in a boolean? Or rethink about how delegatesFocus behaves.
Dan: maybe there is a thing we could think around delegates
Keith: in the presence of a delegatesFocus attribute, maybe that takes precedence, but if you were to provide a string, then it could change that behavior? We could tie it together.
Dan: maybe if you give it an ID, we can change that behavior. There is a question of how this should behave if you also have delegateFocus, but I don’t know that we want to (in a default way) tie these together.

**Comment by EisenbergEffect** [comment]
Ok, so `referenceTarget` is now `delegatesTo`? What about `referenceTargetMap`? Can you also clarify the attribute names for DSD generation purposes? I'd like to update my code...

**Comment by alice** [comment]
@EisenbergEffect My understanding is that `delegatesTo` was somewhat preferred by those at the f2f, but that there hasn't actually been a firm decision to change the name just yet. `referenceTarget` is still what's implemented in Blink and WebKit and afaik there are no issues filed yet to change it, much less code committed.

**Comment by alice** [comment]
My concern with `delegatesTo` is that it's not explicit about what is being delegated. If we're limiting the functionality of this property strictly to delegating references to the host (whether via ID or element reference), which I think we should, I think the name should reflect that. In particular, I think we should leave ourselves room to consider a parallel feature which would allow delegating attributes on the host in a similar way.

**Comment by annevk** [comment]
I don't think we should let what's implemented as a prototype drive the standardization discussion. That seems quite wrong.

I agree that it should be limited to incoming ID or element references, but I don't think the name needs to reflect that explicitly.

I think some kind of naming that's analogous to `delegatesFocus` would be nice, as it's a very similar sort of thing.

**Comment by keithamus** [comment]
> What about `referenceTargetMap`?

I believe we discussed collapsing them into one attribute where a single idref might map to one element (so e.g. `<template shadowrootdelegatesto="inner-el">`) while an expanded syntax would allow for mapping individual properties (`<template shadowrootdelegatesto="inner-label:aria-labelledby inner-p:aria-describedby">`). An altnerative was discussed where we have individual attributes for each (e.g. `<template shadowrootdelegatesarialabelledby="inner-label" shadowrootdelegatesariadescribedby="inner-p">`). This discussion was largely deferred though, as it is considered "phase 2" work - I believe.

**Comment by alice** [comment]
> I don't think we should let what's implemented as a prototype drive the standardization discussion. That seems quite wrong.

Was someone suggesting that we should?

> I agree that it should be limited to incoming ID or element references, but I don't think the name needs to reflect that explicitly.

My concern is that if it's not explicit about what is being delegated, it it sounds like a catch-all, which we all agree it isn't and shouldn't be. We would at the very least need to be explicit with any related attributes, which could only add to the confusion.

For example:

```html
<template shadowrootmode="open"
          shadowrootdelegatesfocus
          shadowrootdelegatesto="input"
          shadowrootdelegatesattributesto="input"> 
```

To me, it seems potentially very confusing: does `shadowrootdelegatesto` refer to focus somehow, since `shadowrootdelegatesfocus` doesn't take an ID? Does it refer to everything (for some unknown value of everything) _other_ than focus and attributes? Is `shadowrootdelegatesattributesto` redundant, since it refers to the same element?

**Comment by rniwa** [comment]
> To me, it seems potentially very confusing: does shadowrootdelegatesto refer to focus somehow, since shadowrootdelegatesfocus doesn't take an ID? Does it refer to everything (for some unknown value of everything) other than focus and attributes? Is shadowrootdelegatesattributesto redundant, since it refers to the same element?

FWIW, I don't think "reference target" is any better in that regard. It's equally confusing because the confusion stems from the semantics of this new attribute, not from its name.

One thought. Focus delegation is delegation of focus. It's a delegation of an action
by user or script. Here, what we are delegating is the ID reference and it's a static/declarative relationship so perhaps there is some distinction we can make in that regard.

**Comment by rniwa** [comment]
> My concern is that if it's not explicit about what is being delegated, it it sounds like a catch-all, which we all agree it isn't and shouldn't be. We would at the very least need to be explicit with any related attributes, which could only add to the confusion.

But it is a catch-all for ID references, no? I thought that was the whole idea behind this phase 1 feature.

**Comment by alice** [comment]
> But it is a catch-all for ID references, no?

Yes, it is a catch-all for ID (and element) references, but not for everything which may be forwarded (for example focus, and attributes on the host, are out of scope).

> FWIW, I don't think "reference target" is any better in that regard.

Reference target is not immediately clear, I agree, but I think it is explicit: it designates a _target_ for _references_ to the host element. I could imagine a suite of similarly-named attributes (`attributeTarget`, etc.)

> Here, what we are delegating is the ID reference and it's a static/declarative relationship so perhaps there is some distinction we can make in that regard.

I think this is a really important point.

**Comment by annevk** [comment]
@Alice I was reacting to something you wrote above that I do not deem relevant to the discussion, but perhaps I misunderstood it:

> referenceTarget is still what's implemented in Blink and WebKit and afaik there are no issues filed yet to change it, much less code committed.

**Comment by bkardell** [comment]
> My understanding is that delegatesTo was somewhat preferred by those at the f2f, but that there hasn't actually been a firm decision to change the name just yet.

This is, for what it is worth, my understanding too. All of the minutes say things like "I don't have strong feelings" etc. 

> FWIW, I don't think "reference target" is any better in that regard. It's equally confusing because the confusion stems from the semantics of this new attribute, not from its name. 

Idk, I personally think it's a bit clearer than `delegatesTo` because there's not a few different kinds of confusion that already use 'delegates' and say something specific. At the same time, I don't think any of the options we've come up with yet is great.  Personally I kind of liked refs because I felt like "if they're all going to be slightly unclear/have to be learned - at least be short", but I can appreciate that others don't like that trade off.

**Comment by dandclark** [comment]
> [@alice](https://github.com/alice) I was reacting to something you wrote above that I do not deem relevant to the discussion, but perhaps I misunderstood it:
> 
> > referenceTarget is still what's implemented in Blink and WebKit and afaik there are no issues filed yet to change it, much less code committed.

@annevk For what it's worth I read it as strictly a reply to Rob's [question](https://github.com/WICG/webcomponents/issues/1087#issuecomment-2673147647) about whether code written against the browser prototypes in anticipation of the feature should be updated at this time. I think the answer to that is still "not yet" given the continued debate here.

Looking at examples like @alice's [here](https://github.com/WICG/webcomponents/issues/1087#issuecomment-2677167922) where it's paired with delegatesFocus does strike me as more confusing than I had anticipated when taking about it in the f2f. This reads to me like the shadow will be delegating focus and it will be delegating it to `"input"`.

```html
<template shadowrootmode="open"
          shadowrootdelegatesfocus
          shadowrootdelegatesto="input"> 
```

And the point is well-taken that we might eventually want to delegate a seperate set of things like attributes, and a generic `delegatesTo` could be a constraint to that.  I'm coming back around to preferring `referencetarget`, or `reftarget`...

**Comment by dandclark** [comment]
This naming question was discussed again at the end of the [June WebEnginesHackfest session](https://github.com/Igalia/webengineshackfest/issues/54), and we resolved to just stick with `referencetarget`. See [notes](https://github.com/Igalia/webengineshackfest/wiki/2025-Cross%E2%80%90shadow-root-references-(referenceTarget)). Perhaps this Issue can be closed?

**Comment by Westbrook** [comment]
Decision: it’s called a `referencetarget`. 

:shipit:
